sql: allow creating a sequence and calling nextval/setval on it within a transaction

ecwall opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/61425

Release note (bug fix): nextval/setval are non-transactional
except when it is called in the same transaction that the sequence
was created in. This change prevents a bug where creating a 
sequence and calling nextval/setval on it within a transaction 
caused the query containing nextval to hang.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77949)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
nice progress!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-:-MyIcL-I8Di-Ceodqc3H:b-o0w08a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyI_Yb24yi9J3Hx-OkV):*
nit: the commit message title should describe what the change is doing. it looks like this title is describing the pre-existing problem

https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages
___
*[-- commits, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIXbns8GMvE4U5xWs0):*
could you make this `Release note (bug fix)` instead? `sql change` would be for changes from one expected behavior to another expected behavior, but in this case there's no way anyone out there is _expecting_ that CRDB would deadlock.
___
*[-- commits, line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIXqJ0-eKhsdYOiMQf):*
can you be more specific? we'll want to capture the output of that slack discussion (https://cockroachlabs.slack.com/archives/C0168LW5THS/p1647376774176739) to say that in most cases `nextval` is non-transactional, but if you call `nextval` in the same transaction as the one that created the sequence, then it is transactional
___
*[pkg/kv/txn.go, line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyI_wOjDcnfpLoqFbwm:-MyI_wOjDcnfpLoqFbwn:bvweyks) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/kv/txn.go#L101)):*
> ```Go
> 	admissionHeader roachpb.AdmissionHeader
> 
> 	CreatedSequenceMap map[uint32]struct{}
> ```

do you know if this has to be stored here? would it work as a field in `extraTxnState`? 

see conn_executor.go. and also take a look at sql_cursors.go to see how you can expose what you need from `extraTxnState`
___
*[pkg/sql/planner.go, line 895 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIaKny7pSo3qA7sELk:-MyIaKny7pSo3qA7sELl:ba3wtn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/planner.go#L895)):*
> ```Go
> }
> 
> func (p *planner) addCreatedSequence(id descpb.ID) {
> ```

this is only used in one place, and doesn't seem like something we'd want other users of a `planner` to call. so i think it would make more sense to inline this code in create_sequence.go. or alternatively, whatever interface you use to expose `extraTxnState` can have this function.

also, please add comments describing something like "keep track of a sequence created in the current transaction"
___
*[pkg/sql/planner.go, line 904 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIbT458iONJkb--Ns3:-MyIbT458iONJkb--Ns4:b-m4q30r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/planner.go#L904)):*
> ```Go
> }
> 
> func (p *planner) isCreatedSequence(id descpb.ID) bool {
> ```

similar to above: either inline or expose on a different interface
___
*[pkg/sql/sequence.go, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIbhIF0gxJYeWG21ZT:-MyIbhIF0gxJYeWG21ZU:b-oijl41) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/sequence.go#L160)):*
> ```Go
> 			// https://www.postgresql.org/docs/14/functions-sequence.html
> 			endValue, err = kv.IncrementValRetryable(
> 				ctx, p.txn.DB(), seqValueKey, seqOpts.Increment*cacheSize)
> ```

it looks like this changed from `p.ExecCfg().DB()` to `p.txn.DB()` -- but i don't think that's what we want, as per the comment right above this line
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2076 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIcDgE2iTaIYkPCJA-:-MyIcDgF34UQqRtq7ti3:bstv6n4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/logictest/testdata/logic_test/sequences#L2076)):*
> ```
>   CACHE 1
> 
> # creating a sequence and calling nextval in the same transaction
> ```

could we add a similar test/fix for `setval`? (pretty sure that is also affected, but i haven't confirmed yet)


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @rafiss, and @RichardJCai)

___
*[pkg/sql/planner.go, line 896 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyJJft121JW-6uEbxFq:-MyJJft121JW-6uEbxFr:bzedafw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/planner.go#L896)):*
> ```Go
> 
> func (p *planner) addCreatedSequence(id descpb.ID) {
> 	createdSequenceMap := p.txn.CreatedSequenceMap
> ```

txn can be accessed concurrently by different goroutines, I think the way it's currently written is susceptible to race conditions.
___
*[pkg/sql/sequence.go, line 139 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyJJyQ_5_jLeJVEImsI:-MyJJyQ_5_jLeJVEImsJ:bn08moo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/sequence.go#L139)):*
> ```Go
> 
> 	sequenceId := descriptor.GetID()
> 	useTxn := p.isCreatedSequence(sequenceId)
> ```

nit: rename `useTxn` to `createdInCurrentTxn`


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @rafiss, and @RichardJCai)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyI_Yb24yi9J3Hx-OkV):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: the commit message title should describe what the change is doing. it looks like this title is describing the pre-existing problem

https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages
</blockquote></details>

Done.
___
*[-- commits, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIXbns8GMvE4U5xWs0):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you make this `Release note (bug fix)` instead? `sql change` would be for changes from one expected behavior to another expected behavior, but in this case there's no way anyone out there is _expecting_ that CRDB would deadlock.
</blockquote></details>

Done.
___
*[-- commits, line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIXqJ0-eKhsdYOiMQf):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you be more specific? we'll want to capture the output of that slack discussion (https://cockroachlabs.slack.com/archives/C0168LW5THS/p1647376774176739) to say that in most cases `nextval` is non-transactional, but if you call `nextval` in the same transaction as the one that created the sequence, then it is transactional
</blockquote></details>

Done.
___
*[pkg/kv/txn.go, line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyI_wOjDcnfpLoqFbwm:-MyJtM88DFy3sz3cTUcm:b-jnptvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/kv/txn.go#L101)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do you know if this has to be stored here? would it work as a field in `extraTxnState`? 

see conn_executor.go. and also take a look at sql_cursors.go to see how you can expose what you need from `extraTxnState`
</blockquote></details>

Using something like this to be able to access it from the `planner`?
```
func (ex *connExecutor) getCursorAccessor() sqlCursors {
	return connExCursorAccessor{
		ex: ex,
	}
}
```
___
*[pkg/sql/planner.go, line 896 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyJJft121JW-6uEbxFq:-MyJur_yBu6npBuf1PIE:b-ttze1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/planner.go#L896)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

txn can be accessed concurrently by different goroutines, I think the way it's currently written is susceptible to race conditions.
</blockquote></details>

Moving this to `extraTxnState`. Does the same apply there?
___
*[pkg/sql/sequence.go, line 139 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyJJyQ_5_jLeJVEImsI:-MyJvr9n3IiHdLCeriKv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/sequence.go#L139)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: rename `useTxn` to `createdInCurrentTxn`
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIbhIF0gxJYeWG21ZT:-MyJvJ-q0DnIKGWLqWVF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/sequence.go#L160)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it looks like this changed from `p.ExecCfg().DB()` to `p.txn.DB()` -- but i don't think that's what we want, as per the comment right above this line
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, @rafiss, and @RichardJCai)

___
*[pkg/kv/txn.go, line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyI_wOjDcnfpLoqFbwm:-MyKFIMsECJy1q86spW9:bv65xn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/kv/txn.go#L101)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Using something like this to be able to access it from the `planner`?
```
func (ex *connExecutor) getCursorAccessor() sqlCursors {
	return connExCursorAccessor{
		ex: ex,
	}
}
```
</blockquote></details>

it shouldn't live on `kv.Txn` as we're now leaking the sequence layer onto `sql`.
yeah you need to do something like what you said to get it working in the planner.
___
*[pkg/sql/sequence.go, line 138 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyKF98-20cQksCs5cXC:-MyKF98-20cQksCs5cXD:bw16tgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3547dee71f332bceb2e378bce4056247bb4e5d83/pkg/sql/sequence.go#L138)):*
> ```Go
> 	seqOpts := descriptor.GetSequenceOpts()
> 
> 	sequenceId := descriptor.GetID()
> ```

nit: acronyms share the same capitalisation, i.e. `sequenceID`


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/planner.go, line 895 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIaKny7pSo3qA7sELk:-MyOZJ5XAlxJL0CZzU6I:b-mnh0as) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/planner.go#L895)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this is only used in one place, and doesn't seem like something we'd want other users of a `planner` to call. so i think it would make more sense to inline this code in create_sequence.go. or alternatively, whatever interface you use to expose `extraTxnState` can have this function.

also, please add comments describing something like "keep track of a sequence created in the current transaction"
</blockquote></details>

Moved to `extraTxnState` and followed the accessor pattern.

Also added a comment.
___
*[pkg/sql/planner.go, line 896 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyJJft121JW-6uEbxFq:-MyO_FTq2e2dQVk2ywQ8:b-28j1jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/planner.go#L896)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Moving this to `extraTxnState`. Does the same apply there?
</blockquote></details>

It looks like this is not accessed concurrently as nothing else is using a lock.
___
*[pkg/sql/planner.go, line 904 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIbT458iONJkb--Ns3:-MyO_MrD5OLLAvH8YDse:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/planner.go#L904)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

similar to above: either inline or expose on a different interface
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 138 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyKF98-20cQksCs5cXC:-MyO_XRx3XxLknM16ygZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3547dee71f332bceb2e378bce4056247bb4e5d83/pkg/sql/sequence.go#L138)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: acronyms share the same capitalisation, i.e. `sequenceID`
</blockquote></details>

Done.



_Code quote:_
```Go
sequenceId := descriptor.GetID()
```
___
*[pkg/kv/txn.go, line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyI_wOjDcnfpLoqFbwm:-MyO_fP0BaLgYcCoINug:bljq4fe) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/kv/txn.go#L101)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

it shouldn't live on `kv.Txn` as we're now leaking the sequence layer onto `sql`.
yeah you need to do something like what you said to get it working in the planner.
</blockquote></details>

Moved to extraTxnState.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2076 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIcDgE2iTaIYkPCJA-:-MyO_mRP-AZVBoDzxo77:b-134fqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/logictest/testdata/logic_test/sequences#L2076)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could we add a similar test/fix for `setval`? (pretty sure that is also affected, but i haven't confirmed yet)
</blockquote></details>

Can this be done in a separate PR?


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2076 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyIcDgE2iTaIYkPCJA-:-Myh6Bjl4MyM6PgBrDzV:bboirs9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/logictest/testdata/logic_test/sequences#L2076)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Can this be done in a separate PR?
</blockquote></details>

Never mind, the change ended up being small so I included it here.


<!-- Sent from Reviewable.io -->

ecwall committed:
sql: allow creating a sequence and calling nextval/setval on it
within a transaction

fixes https://github.com/cockroachdb/cockroach/issues/61425

Release note (bug fix): nextval/setval are non-transactional
except when it is called in the same transaction that the sequence
was created in. This change prevents a bug where creating a
sequence and calling nextval/setval on it within a transaction
caused the query containing nextval to hang.

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/planner.go, line 408 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyseHcL4-z17Po7l2kh:-MyseHcMAJtZ45-_0v3C:b-jo47aq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/planner.go#L408)):*
> ```Go
> 	p.queryCacheSession.Init()
> 	p.optPlanningCtx.init(p)
> 	p.createdSequences = noopCreatedSequences{}
> ```

It seems bad that the `planner` API is "incomplete" unless `connExecutor.initPlanner` is called on the planner since it results in a nil pointer dereference in an unrelated part of the code base if that part of the API is used.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/conn_executor.go, line 1383 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyseOe7A-e-teGY5G3L:-MyseOe7A-e-teGY5G3M:bshhijy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/conn_executor.go#L1383)):*
> ```Go
> 
> 		// keeps track of sequences created in the current transaction
> 		// the map key is the sequence descpb.ID
> ```

super nit, period to end the sentence.
___
*[pkg/sql/planner.go, line 896 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyJJft121JW-6uEbxFq:-MyNLfW_7DcEspCK85Ga:bdftal5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/planner.go#L896)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Moving this to `extraTxnState`. Does the same apply there?
</blockquote></details>

Still an issue, we should use a concurrency safe map and create that upfront somewhere so we don't go around assigning it like `p.txn.CreatedSequenceMap = createdSequenceMap` below
___
*[pkg/sql/planner.go, line 408 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyseHcL4-z17Po7l2kh:-MysgTSo1IT2uxLgaeRv:b-i29td1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/planner.go#L408)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

It seems bad that the `planner` API is "incomplete" unless `connExecutor.initPlanner` is called on the planner since it results in a nil pointer dereference in an unrelated part of the code base if that part of the API is used.
</blockquote></details>

I think it should probably be right that `initPlanner` should always be called before `p.createdSequences` is used (ie if we create a sequence) - at least I hope that's the case.

If we want to keep this `noopCreatedSequences` thing around, I think it's functions should probably panic, also can we make a test where we create an `InternalPlanner` and try creating a sequence. We can keep `noopCreatedSequences` around to use for assertions?

I think `p.createdSequences` should actually be assigned in `initPlanner`
___
*[pkg/sql/sql_cursor.go, line 287 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-Mysejeq014g4UtYJi4E:-Mysejeq014g4UtYJi4F:b-90kmfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/sql_cursor.go#L287)):*
> ```Go
> 
> type createdSequences interface {
> 	// adds a sequence to the set of sequences created in the current transaction
> ```

nit: end with period.
___
*[pkg/sql/sql_cursor.go, line 289 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MysekYq2HvvIj5wSHgz:-MysekYq2HvvIj5wSHh-:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/sql_cursor.go#L289)):*
> ```Go
> 	// adds a sequence to the set of sequences created in the current transaction
> 	addCreatedSequence(id uint32)
> 	// checks if a sequence was created in the current transaction
> ```

ditto
___
*[pkg/sql/sql_cursor.go, line 293 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MysectU6-_qe_DWev1N:-MysectU6-_qe_DWev1O:b3pjo0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/sql_cursor.go#L293)):*
> ```Go
> }
> 
> type noopCreatedSequences struct {
> ```

nit: prefer open and close brace on same line in this case, and below.

Also can we add a comment on why this is needed?


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/conn_executor.go, line 1383 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyseOe7A-e-teGY5G3L:-MywbDcf3BeJ1eqMTSBd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/conn_executor.go#L1383)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

super nit, period to end the sentence.
</blockquote></details>

Done.
___
*[pkg/sql/planner.go, line 408 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyseHcL4-z17Po7l2kh:-MywAO37Ck_os9Ka2HAQ:bp0kb0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/planner.go#L408)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I think it should probably be right that `initPlanner` should always be called before `p.createdSequences` is used (ie if we create a sequence) - at least I hope that's the case.

If we want to keep this `noopCreatedSequences` thing around, I think it's functions should probably panic, also can we make a test where we create an `InternalPlanner` and try creating a sequence. We can keep `noopCreatedSequences` around to use for assertions?

I think `p.createdSequences` should actually be assigned in `initPlanner`
</blockquote></details>

I only created `noopCreatedSequences` to get tests to pass, but it seems like a bad design.

What is the best way to access the `connExecutor` from here?
___
*[pkg/sql/sql_cursor.go, line 287 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-Mysejeq014g4UtYJi4E:-MywaPuWADy_qQflgiU0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/sql_cursor.go#L287)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: end with period.
</blockquote></details>

Done.
___
*[pkg/sql/sql_cursor.go, line 289 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MysekYq2HvvIj5wSHgz:-MywaQnj2QoIJqRyxtwJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/sql_cursor.go#L289)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/sql/sql_cursor.go, line 293 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MysectU6-_qe_DWev1N:-MywaSIX6KHBLgJ31uDf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06bdf0214ccf4c0826734457fe2d5687a3d733a/pkg/sql/sql_cursor.go#L293)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: prefer open and close brace on same line in this case, and below.

Also can we add a comment on why this is needed?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/planner.go, line 896 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MyJJft121JW-6uEbxFq:-Mz0Grgd38HtIkFOxF4F:br9z55z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b79bf4c7a77d2979cfdaceda5f8eadf9f5ac53e9/pkg/sql/planner.go#L896)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Still an issue, we should use a concurrency safe map and create that upfront somewhere so we don't go around assigning it like `p.txn.CreatedSequenceMap = createdSequenceMap` below
</blockquote></details>

I'm not sure it correct to use this concurrently even if we made the map thread safe. Concurrent access would lead to different results depending on which thread got there first.

Don't the expressions in a statement need to be evaluated one at a time in a specific order?


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
just some small nits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-:-MzLOhMl5O8bRYP_jQuz:b-xwvllc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/conn_executor.go, line 1385 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzLMwt7131CkAFwBLAu:-MzLMwt7131CkAFwBLAv:baeswpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a551d475659531ca5155a45d44ec8c533f1df41/pkg/sql/conn_executor.go#L1385)):*
> ```Go
> 		// createdSequences keeps track of sequences created in the current transaction.
> 		// The map key is the sequence descpb.ID.
> 		createdSequences map[uint32]struct{}
> ```

nit: we should be able to use `descpb.ID` as the map key type
___
*[pkg/sql/conn_executor.go, line 1655 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzLMaSI8ncP36XtxA9g:-MzLMaSI8ncP36XtxA9h:bqdp3il) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a551d475659531ca5155a45d44ec8c533f1df41/pkg/sql/conn_executor.go#L1655)):*
> ```Go
> 	ex.extraTxnState.sqlCursors.closeAll()
> 
> 	createdSequences := ex.extraTxnState.createdSequences
> ```

nit: we can just initialize a new map:
```
ex.extraTxnState.createdSequences = make(map...)
```
___
*[pkg/sql/created_sequence.go, line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzLNbY43Aj_NDTvf7sM:-MzLNbY43Aj_NDTvf7sN:b-u7uza0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a551d475659531ca5155a45d44ec8c533f1df41/pkg/sql/created_sequence.go#L37)):*
> ```Go
> 
> func (createdSequences noopCreatedSequences) addCreatedSequence(id uint32) {
> 	panic("addCreatedSequence not supported")
> ```

this has the potential to crash the node, so i think it would be better to return `errors.AssertionFailed`
___
*[pkg/sql/sequence.go, line 336 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzLOUElBZpD-vKv2YIo:-MzLOUElBZpD-vKv2YIp:b9bm3ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a551d475659531ca5155a45d44ec8c533f1df41/pkg/sql/sequence.go#L336)):*
> ```Go
> 	}
> 
> 	// We *do not* use the planner txn here, since setval does not respect
> ```

i think this comment should be modified to explain the part right above, where sometimes we do use the planner txn


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/conn_executor.go, line 1385 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzLMwt7131CkAFwBLAu:-MzLuo4jC3CgzrDeqFj8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a551d475659531ca5155a45d44ec8c533f1df41/pkg/sql/conn_executor.go#L1385)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we should be able to use `descpb.ID` as the map key type
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 1655 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzLMaSI8ncP36XtxA9g:-MzLupwl2h6qZDCIaOdH:bflsdna) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a551d475659531ca5155a45d44ec8c533f1df41/pkg/sql/conn_executor.go#L1655)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we can just initialize a new map:
```
ex.extraTxnState.createdSequences = make(map...)
```
</blockquote></details>

I'll change it but it looks like the other maps here are having their entries deleted.

Should those be changed as well for consistency?
___
*[pkg/sql/created_sequence.go, line 37 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzLNbY43Aj_NDTvf7sM:-MzLvIp73jUrLaBnV4hU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a551d475659531ca5155a45d44ec8c533f1df41/pkg/sql/created_sequence.go#L37)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this has the potential to crash the node, so i think it would be better to return `errors.AssertionFailed`
</blockquote></details>

Done.
___
*[pkg/sql/sequence.go, line 336 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzLOUElBZpD-vKv2YIo:-MzLzRFq3znI5wrNzxxA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a551d475659531ca5155a45d44ec8c533f1df41/pkg/sql/sequence.go#L336)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think this comment should be modified to explain the part right above, where sometimes we do use the planner txn
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (approved):
LGTM modulo last comment

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-:-MzSD0dJ5upmc3rnvVqv:b-nlug88)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/created_sequence.go, line 40 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzSCJ30BWocyjja5HL4:-MzSCJ30BWocyjja5HL5:b4zm4wz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a724e60aba9b3e144e4787d1da2944b41d9bd5c/pkg/sql/created_sequence.go#L40)):*
> ```Go
> 
> // noopCreatedSequences is the default impl used by the planner when the connExecutor is not available.
> type noopCreatedSequences struct{}
> ```

nit: IMO `emptyCreatedSequencesMap` is more appropriate since the noop really applies to `isCreatedSequence` which we can leave a comment there about.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2104 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzSCwpe9eCf5EV5RxGU:-MzSCwpe9eCf5EV5RxGV:b-okck1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a724e60aba9b3e144e4787d1da2944b41d9bd5c/pkg/sql/logictest/testdata/logic_test/sequences#L2104)):*
> ```
> 
> query I
> SELECT setval('setval_txn_seq', 2)
> ```

Can we add another check in this transaction doing nextval and see if that increments correctly?


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @RichardJCai)

___
*[pkg/sql/created_sequence.go, line 40 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzSCJ30BWocyjja5HL4:-MzubVwCAvJR1c82whsv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a724e60aba9b3e144e4787d1da2944b41d9bd5c/pkg/sql/created_sequence.go#L40)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: IMO `emptyCreatedSequencesMap` is more appropriate since the noop really applies to `isCreatedSequence` which we can leave a comment there about.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/sequences, line 2104 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/77949#-MzSCwpe9eCf5EV5RxGU:-MzpeXVaAIxTRnWbf9nt:b-kaun3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a724e60aba9b3e144e4787d1da2944b41d9bd5c/pkg/sql/logictest/testdata/logic_test/sequences#L2104)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Can we add another check in this transaction doing nextval and see if that increments correctly?
</blockquote></details>

It looks like we still have some failing so I created tickets and test cases for them.


<!-- Sent from Reviewable.io -->

ecwall commented:
bors r=RichardJCai

ecwall commented:
blathers backport 21.1 21.2 22.1
